### ISSUE TITLE
Setting up CI/CD

### ISSUE THREAD

**Issue by trezy** [Ariel-Thomas/adventurers-league-log#145]
## TL;DR

I think we should set up CI/CD. @Ariel-Thomas, you're obviously welcome to do it yourself. For me to dig into it, though, I'd have to be added as a contributor on the repo and added to the team on Heroku. I can get the CircleCI stuff set up from scratch and add you to the team.

## The long version

As far as I'm aware, we don't have any form of CI/CD set up at the moment. I'd be happy to dive in and make that a reality to shorten the time it takes for features to go from development to production. Here's the setup I usually uses, adapted for the way that I *think* our infrastructure is currently set up:

### Continuous Integration via CircleCI

[CircleCI](https://circleci.com/) has a super robust framework for CI that I absolutely adore. I usually set up CircleCI to run builds and tests on every commit to the website. This gets status checks showing up on PRs and such, which means we don't have to run the tests manually before merging a PR.

### Continuous Deployment with CircleCI and Heroku

I think we already deploy to Heroku. If that's the case, CircleCI and Heroku make deployments really simple. When anything gets merged to the `develop` branch, it gets deployed automatically to the development website (https://dev.adventurersleaguelog.com?), and the same thing happens for the `master` branch and the production website. The next section details how to make sure that automated deployments are still safe, though.

### Locking down the repo

CI/CD is awesome, but with the amount of automation I tend to build in, it's not the best idea to leave it open. Here's my standard repo set up:

#### Feature branches

Most of our work happens on feature branches, so this isn't anything new. The big difference is that if we move forward with the rest of these changes, it'll be an *absolute necessity* that all of our work happens on feature branches, as we won't be able to push directly to `develop` or `master`.

#### `develop` branch

The `develop` branch is the source of truth for the development website. It's a bit more wild west than the `master` branch, but for the most part the `develop` branch should almost always be in a deployable state. This is where we start really locking things down, though.

I always [set up `develop` as a "protected" branch](https://github.com/Ariel-Thomas/adventurers-league-log/settings/branches/develop). [These are the settings I use](https://trezy.sh/ZcQ7m0.png).

1. Restricting who can push to the branch prevents anybody from being able to push directly to it. This mostly happens when people pull `develop` and forget to make a new branch before making their changes.
1. Requiring status checks to pass means that the tests will have to pass before we merge. That doesn't mean nothing broken will make it though, but it at least means that the parts of the site that have tests should be in the clear.
1. The purpose of requiring PR reviews is similar to restricting who can push to the branch. It prevents anybody from pushing a feature branch, creating a PR, and merging it without any oversight.

#### `master` branch

The `master` branch is the source of truth for the production website. It takes on all of the same properties of the `develop` branch for safety's sake. The biggest difference is that I actually pull down the code and make sure nothing's obviously broken before approving a PR for `master`. On `develop`, I just read through the code and check for code smells and obvious issues.

**Comment by Ariel-Thomas** [comment]
I actually do have CI via Codeship, it's just not publicly visible.

I'll think about something more public and your suggestions on the branches and such.

**Comment by whazzmaster** [comment]
I'm also happy to assist on any infrastructure tasks.
